impeller/renderer/backend/vulkan/fence_waiter_vk.cc

```diff
@@ -101,76 +95,90 @@ void FenceWaiterVK::Main() {
+  if (!device_holder) {
+    return false;
+  }
 
-    // One or more fences have been signaled. Find out which ones and update
-    // their signaled statuses.
-    {
-      TRACE_EVENT0("impeller", "CheckFenceStatus");
-      for (auto& entry : wait_set) {
-        entry->UpdateSignalledStatus(device);
-      }
-      wait_set.clear();
-    }
+  const auto& device = device_holder->GetDevice();
+  // Wait for one or more fences to be signaled. Any additional fences added
+  // to the waiter will be serviced in the next pass. If a fence that is going
+  // to be signaled at an abnormally long deadline is the only one in the set,
+  // a timeout will bail out the wait.
+  auto fences = GetFencesForWaitSet(wait_set);
+  if (fences.empty()) {
+    return true;
+  }
 
-    // Quickly acquire the wait set lock and erase signaled entries. Make sure
-    // the mutex is unlocked before calling the destructors of the erased
-    // entries. These might touch allocators.
-    WaitSet erased_entries;
-    {
-      static auto is_signalled = [](const auto& entry) {
-        return entry->IsSignalled();
-      };
-      std::scoped_lock lock(wait_set_mutex_);
-      std::copy_if(wait_set_.begin(), wait_set_.end(),
-                   std::back_inserter(erased_entries), is_signalled);
-      wait_set_.erase(
-          std::remove_if(wait_set_.begin(), wait_set_.end(), is_signalled),
-          wait_set_.end());
-    }
+  auto result = device.waitForFences(
+      fences.size(),                           // fences count
+      fences.data(),                           // fences
+      false,                                   // wait for all
+      std::chrono::nanoseconds{100ms}.count()  // timeout (ns)
+  );
+  if (!(result == vk::Result::eSuccess || result == vk::Result::eTimeout)) {
+    VALIDATION_LOG << "Fence waiter encountered an unexpected error. Tearing "
+                      "down the waiter thread.";
+    return false;
+  }
 
-    {
-      TRACE_EVENT0("impeller", "ClearSignaledFences");
-      // Erase the erased entries which will invoke callbacks.
-      erased_entries.clear();  // Bit redundant because of scope but hey.
+  // One or more fences have been signaled. Find out which ones and update
+  // their signaled statuses.
+  {
+    TRACE_EVENT0("impeller", "CheckFenceStatus");
+    for (auto& entry : wait_set) {
+      entry->UpdateSignalledStatus(device);
     }
+    wait_set.clear();
+  }
+
+  // Quickly acquire the wait set lock and erase signaled entries. Make sure
+  // the mutex is unlocked before calling the destructors of the erased
+  // entries. These might touch allocators.
+  WaitSet erased_entries;
+  {
+    static auto is_signalled = [](const auto& entry) {
+      return entry->IsSignalled();
+    };
+    std::scoped_lock lock(wait_set_mutex_);
+    std::copy_if(wait_set_.begin(), wait_set_.end(),
+                 std::back_inserter(erased_entries), is_signalled);
+    wait_set_.erase(
```

Comment:
I'll add a TODO here, not because I don't agree but there is a higher chance I'll make a mistake than improve anything and this unblocks several other folks/initiatives.